Add P2SH-P2WSH support to listunspent RPC

meshcollider opened the pull request:
This is a reworked version of #11708 after #12427 and the `signrawtransaction` split.

For a P2WSH address, listunspent should return the witness script, and for a P2SH-P2WSH address, it should also return the inner witness script (because SignTransaction will automatically wrap it in P2SH if required).

Includes a test which also tests the behaviour of #12427, and release note.

meshcollider committed:
Make listunspent and signrawtransaction RPCs support witnessScript

meshcollider committed:
Add test for P2SH-P2WSH in signrawtransactionwithkey and listunspent

meshcollider committed:
Add release note for listunspent P2WSH change

Sjors commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2709,7 +2709,7 @@ static UniValue listunspent(const JSONRPCRequest& request)
             "    \"amount\" : x.xxx,         (numeric) the transaction output amount in " + CURRENCY_UNIT + "\n"
             "    \"confirmations\" : n,      (numeric) The number of confirmations\n"
-            "    \"redeemScript\" : n        (string) The redeemScript if scriptPubKey is P2SH\n"
+            "    \"redeemScript\" : n        (string) The redeemScript if scriptPubKey is P2SH, or witnessScript if the scriptPubKey is P2WSH or P2SH-P2WSH\n"
```
Why not a separate `witnessScript` field like in #14454?

MeshCollider replied:
@Sjors because this has to be compatible with how signrawtransaction works, so it can be passed directly in. signrawtransaction only accepts the witness script and automatically wraps it

Sjors replied:
Would it be super involved to add support for witnessScript to `signrawtransactionwithkey`?

instagibbs replied:
Agreed with @Sjors I'd much rather we fix terminology instead of compounding confusion.

ryanofsky reviewed (approved):
Code utACK c789fef8ca4739e2fb3c930bacad77543650b781, but I agree with others about not wanting to overload redeemScript in `listunspent` https://github.com/bitcoin/bitcoin/pull/14481#discussion_r225830355. It seems like it should be possible to be compatible with signraw without overloading.

sipa commented:
I wonder if it isn't just easier to support passing in a descriptor in `signrawtransactionwithkey` and using #14477 to get to get descriptors from listunspent etc? The various ways of passing around scripts in different RPCs looks like hard to solve mess.

ryanofsky commented:
@MeshCollider is there an update on the status of this PR?

c789fef8ca4739e2fb3c930bacad77543650b781 seems like an improvement to me, and has my ACK. But I do think it would be an easy win to replace "redeemScript" with "witnessScript" there, and tweak `signrawtransaction` to accept "witnessScript", so all RPCs could use these terms consistently.

Or if we want to abandon this approach and jump to descriptors instead as suggested in https://github.com/bitcoin/bitcoin/pull/14481#issuecomment-438861892, this seems like another good option. Do you have thoughts on it? Does it require more work? Is it blocked on anything now that #14477 is merged?

MeshCollider commented:
@ryanofsky I plan on modifying this for terminology as above for now, and then I think we should look at RPC descriptor support as a whole after that. But I will come back to this PR in the next few days

MeshCollider commented:
I've rebased this and addressed the terminology comments above, moving to a new witnessScript field with backwards compatibility

Sjors reviewed (commented):
bd2b1f65e949a1768c6f989677ea24b746384ca6 looks OK, but...

I have my doubt if backwards compatibility with something that was already buggy is worth the confusion of overloading `redeemScript` for `listunspent`. I'd rather have it `null` if `signrawtransaction` can infer it anyway, or even just the P2SH wrapper redeem script.

It might be better to squash the first and last commit.

I think `addmultisigaddress` should also return both `redeemScript` and `witnessScript` in the correct manner, so you you can just test:

```python
assert_equal(unspent_output["witnessScript"], p2sh_p2wsh_address["witnessScript"])
assert_equal(unspent_output["redeemScript"], p2sh_p2wsh_address["redeemScript"])
assert_equal(unspent_output["redeemScript"], CScript([OP_0, sha256(hex_str_to_bytes(unspent_output["witnessScript"])))
```

MeshCollider commented:
Forgot to update the release notes, listunspent doesn't overload redeemScript anymore

MeshCollider commented:
I've squashed and updated the release notes, but modifying `addmultisigaddress` to return a witnessScript as well as redeemScript would be a breaking change, because currently the witnessScript is returned as "redeemScript". It would be more confusing to use different terminology so unless a breaking change is worth it here, which I don't think it is, I'll leave it as is

laanwj commented on doc/release-notes-14481.md:
```diff
@@ -0,0 +1,9 @@
+the witness script in the case of a P2WSH or P2SH-P2WSH output.
+
+The `signrawtransactionwithkey` and `signrawtransactionwithwallet` RPCs have been
+modified so that they also accept a `witnessScript`, the witness script in the
```
maybe `optionally` instead of `also`

laanwj commented on test/functional/rpc_signrawtransaction.py:
```diff
@@ -143,9 +145,33 @@ def script_verification_error_test(self):
+        p2sh_redeemScript = CScript([OP_0, sha256(hex_str_to_bytes(p2sh_p2wsh_address["redeemScript"]))])
+        assert_equal(unspent_output["redeemScript"], bytes_to_hex_str(p2sh_redeemScript))
+        # Now create and sign a transaction spending that output on node[0], which doesn't know the scripts or keys
+        spending_tx = self.nodes[0].createrawtransaction([unspent_output], {self.nodes[1].getnewaddress(): 49.998})
```
Please pass fractional amounts as a `Decimal`

Sjors commented on src/rpc/rawtransaction.cpp:
```diff
@@ -852,15 +852,25 @@ UniValue SignTransaction(interfaces::Chain& chain, CMutableTransaction& mtx, con
                     CScript redeemScript(rsData.begin(), rsData.end());
                     keystore->AddCScript(redeemScript);
                     // Automatically also add the P2WSH wrapped version of the script (to deal with P2SH-P2WSH).
+                    // This is only for compatibility, it is encouraged to use the explicit witnessScript field instead.
```
I was also referring to this overloading.

Sjors commented:
> to return a witnessScript as well as redeemScript would be a breaking change, because currently the witnessScript is returned as "redeemScript"

I could be wrong, but I suspect very few people use `addmultisigaddress` with SegWit in an automated setup (maybe @alecalve has stats on how much SegWit multisig is used in general). So it might still be early enough for a breaking change to use `redeemScript` and `witnessScript` accurately and consistently throughout the RPC.

Sjors reviewed (commented):
> to return a witnessScript as well as redeemScript would be a breaking change, because currently the witnessScript is returned as "redeemScript"

I could be wrong, but I suspect very few people use `addmultisigaddress` with SegWit in an automated setup (maybe @alecalve has stats on how much SegWit multisig is used in general). So it might still be early enough for a breaking change to use `redeemScript` and `witnessScript` accurately and consistently throughout the RPC.

(I can live with not doing it though)

ryanofsky commented on src/rpc/rawtransaction.cpp:
```diff
@@ -945,7 +955,8 @@ static UniValue signrawtransactionwithkey(const JSONRPCRequest& request)
                                     {"vout", RPCArg::Type::NUM, /* opt */ false, /* default_val */ "", "The output number"},
                                     {"scriptPubKey", RPCArg::Type::STR_HEX, /* opt */ false, /* default_val */ "", "script key"},
-                                    {"redeemScript", RPCArg::Type::STR_HEX, /* opt */ true, /* default_val */ "omitted", "(required for P2SH or P2WSH) redeem script"},
+                                    {"redeemScript", RPCArg::Type::STR_HEX, /* opt */ true, /* default_val */ "omitted", "(required for P2SH) redeem script"},
```
In commit "Make listunspent return witness script for P2WSH or P2SH-P2WSH" (0b665ab577ab41e3b7b12d9b7d83c151cf90fdaf)

Could update commit message to mention change to signrawtransactionwithkey

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2876,6 +2877,27 @@ static UniValue listunspent(const JSONRPCRequest& request)
+                    // Now check if the redeemScript is actually a P2WSH script
+                    CTxDestination witness_destination;
+                    if (redeemScript.IsPayToWitnessScriptHash()) {
+                        assert(ExtractDestination(redeemScript, witness_destination));
```
In commit "Make listunspent return witness script for P2WSH or P2SH-P2WSH" (0b665ab577ab41e3b7b12d9b7d83c151cf90fdaf)

This would be broken if compiled with NDEBUG, which even though we don't support it, isn't really great for readability because you don't expect to see code with side effects inside an assert. Would be a little better to write something like:

```c++
bool extracted = ExtractDestitionation...;
assert(extracted);
```

ryanofsky reviewed (approved):
utACK 77c0c8e246e1a7f8c403c19d13c8bb699dba8cf4. Left some comments but I think this is good to merge as-is. Only changes since last review are rebase and following up on previous suggestions (main one being to switch from redeemScript to witnessScript). Thanks for adopting these!

> change to use `redeemScript` and `witnessScript` accurately and consistently throughout the RPC

Would agree with Sjors in preferring to use these terms consistently, and breaking backwards compatibility in what seems to be a minor edge case. This could be done in a followup PR, and would be easy thing to explain in release notes.

MarcoFalke commented:
Needs rebase (blame me)

MeshCollider commented:
Rebased and addressed comments above. @MarcoFalke I expect a review from you then ;)

MarcoFalke commented:
travis failure:

```
./test/functional/rpc_signrawtransaction.py:165:108: F821 undefined name 'Decimal'

ryanofsky reviewed (approved):
utACK 6ca836ab3abef5a90df0c3c4e4983f328b1afe00. Changes since last review: rebase due to #14918 and making various requested changes (updating release notes and first commit message, removing assert side-effect, using python Decimal).

MarcoFalke commented:
utACK 6ca836ab3abef5a90df0c3c4e4983f328b1afe00
